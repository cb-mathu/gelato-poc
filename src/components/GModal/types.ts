import { ExtractPropTypes, AllowedComponentProps } from 'vue';
import { MODAL_PROPS } from './constants';

export type ModalProps = ExtractPropTypes<typeof MODAL_PROPS> & AllowedComponentProps;
