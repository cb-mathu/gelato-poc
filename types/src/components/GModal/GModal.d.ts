import { ModalProps } from './types';
declare const _default: new () => {
    $props: ModalProps;
};
export default _default;
