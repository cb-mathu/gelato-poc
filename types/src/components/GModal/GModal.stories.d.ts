import { ModalProps } from './types';
declare const _default: {
    title: string;
    component: new () => {
        $props: ModalProps;
    };
};
export default _default;
export declare const Story: any;
export declare const DefaultModal: (props?: ModalProps | undefined) => JSX.Element;
export declare const ModalWithDescription: () => JSX.Element;
export declare const ModalWithContent: () => JSX.Element;
export declare const ModalWithFooter: () => JSX.Element;
