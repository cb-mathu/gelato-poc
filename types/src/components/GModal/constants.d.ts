import { PropType } from 'vue';
export declare const MODAL_PROPS: {
    isOpen: {
        type: BooleanConstructor;
        required: boolean;
    };
    onClose: {
        type: PropType<() => void>;
        required: boolean;
    };
    title: {
        type: StringConstructor;
        required: boolean;
    };
    description: {
        type: StringConstructor;
        required: boolean;
    };
};
