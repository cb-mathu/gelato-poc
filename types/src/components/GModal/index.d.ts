import GModal from './GModal';
export { GModal };
