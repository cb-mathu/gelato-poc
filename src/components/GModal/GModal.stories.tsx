import GModal from './GModal';
import { ModalProps } from './types';
import { ButtonVariant, GButton } from '../../components/GButton';

export default {
  title: 'Component/GModal',
  component: GModal,
};

const onCloseHandler = () => console.log('onClose is called');

const slots = {
  content: () => (
    <div>
      <p>
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text
        ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived
        not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the
        1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like
        Aldus PageMaker including versions of Lorem Ipsum.
      </p>
      <p>
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text
        ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived
        not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the
        1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like
        Aldus PageMaker including versions of Lorem Ipsum.
      </p>
      <p>
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text
        ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived
        not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the
        1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like
        Aldus PageMaker including versions of Lorem Ipsum.
      </p>
      <p>
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text
        ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived
        not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the
        1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like
        Aldus PageMaker including versions of Lorem Ipsum.
      </p>
    </div>
  ),
  footer: () => <GButton>Submit</GButton>,
};

const multiActionsFooter = () => (
  <>
    <GButton>Secondary Action</GButton>
    <GButton variant={ButtonVariant.Primary}>Primary Action</GButton>
  </>
);

const Template = (args: any) => ({
  setup() {
    return () => <GModal {...args} onClose={onCloseHandler} v-slots={slots} />;
  },
});

export const Story: any = Template.bind({});
Story.args = {
  isOpen: true,
  title: 'Terms of service',
  description:
    "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took",
};

export const DefaultModal = (props?: ModalProps) => {
  const onClose = props?.onClose || onCloseHandler;
  return (
    <GModal
      isOpen
      title="Terms of service"
      description="Lorem Ipsum is simply dummy text of the printing and typesetting industry."
      v-slots={slots}
      {...props}
      onClose={onClose}
    />
  );
};

export const ModalWithDescription = () => (
  <GModal
    title="Terms of service"
    description="Lorem Ipsum is simply dummy text of the printing and typesetting industry."
    isOpen
    onClose={onCloseHandler}
  />
);

export const ModalWithContent = () => (
  <GModal title="Terms of service" onClose={onCloseHandler} isOpen v-slots={{ content: slots.content }} />
);

export const ModalWithFooter = () => (
  <GModal
    title="Terms of service"
    description="Lorem Ipsum is simply dummy text of the printing and typesetting industry."
    isOpen
    onClose={onCloseHandler}
    v-slots={{ footer: multiActionsFooter }}
  />
);
