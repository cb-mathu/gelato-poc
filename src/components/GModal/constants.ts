import { PropType } from 'vue';

export const MODAL_PROPS = {
  isOpen: {
    type: Boolean,
    required: true,
  },
  onClose: {
    type: Function as PropType<() => void>,
    required: true,
  },
  title: {
    type: String,
    required: true,
  },
  description: {
    type: String,
    required: false,
  },
};
