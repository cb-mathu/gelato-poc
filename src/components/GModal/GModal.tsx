import { Dialog, DialogOverlay, DialogTitle, DialogDescription } from '@headlessui/vue';
import { isNotEmpty } from '../../utils/common';
import { defineComponent } from 'vue';
import { Close } from '../../components/icons';
import { MODAL_PROPS } from './constants';
import { ModalProps } from './types';

const GModal = defineComponent({
  name: 'GModal',
  props: MODAL_PROPS,

  setup(props, { slots }) {
    return () => {
      if (!props.isOpen) {
        return null;
      }
      // https://github.com/vuejs/jsx-next/issues/134#issuecomment-721030287
      const customEventListeners = {
        onClose: props.onClose,
      };

      const Title = () => <DialogTitle class="g-modal__title">{props.title}</DialogTitle>;

      const Description = () => (props.description ? <DialogDescription>{props.description}</DialogDescription> : null);

      const Content = () => (isNotEmpty(slots?.content?.()) ? <div class="g-modal__body">{slots.content?.()}</div> : null);

      const Footer = () => (isNotEmpty(slots?.footer?.()) ? <div class="g-modal__footer">{slots?.footer?.()}</div> : null);

      return (
        <Dialog open={props.isOpen} static class="g-modal" {...customEventListeners}>
          <div class="g-modal__container">
            <DialogOverlay class="g-modal__overlay" />
            {/* Dummy button for trapping focus - headlessui's Modal component sets focus to the first focusable element
            in the DOM of the dialog. So if the content is large & is scrollable, and we have an action button at the bottom,
            it'll focused meaning the modal content will be scrolled to the bottom. This doesn't work even with `initialFocus`
            prop that lets you set focus to an arbitrary element within the dialog */}
            <button class="g-modal__test"></button>
            <div class="g-modal__content">
              <Title />
              <Description />
              <Content />
              <Footer />
            </div>
            <button type="button" aria-label="Close" class="g-modal__close" onClick={props.onClose}>
              <Close />
            </button>
          </div>
        </Dialog>
      );
    };
  },
});

export default GModal as {
  new (): {
    $props: ModalProps;
  };
};
