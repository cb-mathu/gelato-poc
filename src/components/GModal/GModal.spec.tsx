import { render, screen, fireEvent } from '@testing-library/vue';
import * as stories from './GModal.stories';

const { DefaultModal, ModalWithContent, ModalWithDescription, ModalWithFooter } = stories;

function getPortalRoot() {
  return document.getElementById('headlessui-portal-root')!;
}

describe('GModal : ', () => {
  test('renders modal by default with any failures', () => {
    render(DefaultModal());
    expect(getPortalRoot().firstChild).toMatchSnapshot();
  });

  test('renders modal with Descritpion', () => {
    render(ModalWithDescription());
    expect(getPortalRoot().firstChild).toMatchSnapshot();
  });

  test('renders modal with Footer', () => {
    render(ModalWithFooter());
    expect(getPortalRoot().firstChild).toMatchSnapshot();
  });

  test('renders modal with Content', () => {
    render(ModalWithContent());
    expect(getPortalRoot().firstChild).toMatchSnapshot();
  });

  test('fires close event on clicking close modal', () => {
    const handleClick = jest.fn();
    render(DefaultModal({ onClose: handleClick, isOpen: true }));
    fireEvent.click(screen.getByLabelText('Close'));
    expect(handleClick).toHaveBeenCalled();
  });

  test('does not render when isOpen is false', () => {
    render(DefaultModal({ isOpen: false }));
    expect(getPortalRoot()).toBe(null);
  });
});
