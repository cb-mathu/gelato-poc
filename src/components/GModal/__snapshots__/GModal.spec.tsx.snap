// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GModal :  renders modal by default with any failures 1`] = `
<div>
  <div
    aria-modal="true"
    class="g-modal"
    id="headlessui-dialog-1"
    role="dialog"
  >
    <div
      class="g-modal__container"
    >
      <div
        aria-hidden="true"
        class="g-modal__overlay"
        id="headlessui-dialog-overlay-2"
      />
      <button
        class="g-modal__test"
      />
      <div
        class="g-modal__content"
      >
        <h2
          class="g-modal__title"
          id="headlessui-dialog-title-3"
        >
          Terms of service
        </h2>
        <p
          id="headlessui-description-4"
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry.
        </p>
        <div
          class="g-modal__body"
        >
          
          <div>
            <p>
              Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
            </p>
            <p>
              Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
            </p>
            <p>
              Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
            </p>
            <p>
              Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
            </p>
          </div>
          
        </div>
        <div
          class="g-modal__footer"
        >
          
          <button
            as="button"
            class="g-button"
            fluid="false"
            loading="false"
          >
            <!---->
            <span
              class="g-button__label"
            >
              
              Submit
              
            </span>
          </button>
          
        </div>
      </div>
      <button
        aria-label="Close"
        class="g-modal__close"
        type="button"
      >
        <svg
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`GModal :  renders modal with Content 1`] = `
<div>
  <div
    aria-modal="true"
    class="g-modal"
    id="headlessui-dialog-13"
    role="dialog"
  >
    <div
      class="g-modal__container"
    >
      <div
        aria-hidden="true"
        class="g-modal__overlay"
        id="headlessui-dialog-overlay-14"
      />
      <button
        class="g-modal__test"
      />
      <div
        class="g-modal__content"
      >
        <h2
          class="g-modal__title"
          id="headlessui-dialog-title-15"
        >
          Terms of service
        </h2>
        <!---->
        <div
          class="g-modal__body"
        >
          
          <div>
            <p>
              Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
            </p>
            <p>
              Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
            </p>
            <p>
              Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
            </p>
            <p>
              Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
            </p>
          </div>
          
        </div>
        <!---->
      </div>
      <button
        aria-label="Close"
        class="g-modal__close"
        type="button"
      >
        <svg
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`GModal :  renders modal with Descritpion 1`] = `
<div>
  <div
    aria-modal="true"
    class="g-modal"
    id="headlessui-dialog-5"
    role="dialog"
  >
    <div
      class="g-modal__container"
    >
      <div
        aria-hidden="true"
        class="g-modal__overlay"
        id="headlessui-dialog-overlay-6"
      />
      <button
        class="g-modal__test"
      />
      <div
        class="g-modal__content"
      >
        <h2
          class="g-modal__title"
          id="headlessui-dialog-title-7"
        >
          Terms of service
        </h2>
        <p
          id="headlessui-description-8"
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry.
        </p>
        <!---->
        <!---->
      </div>
      <button
        aria-label="Close"
        class="g-modal__close"
        type="button"
      >
        <svg
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`GModal :  renders modal with Footer 1`] = `
<div>
  <div
    aria-modal="true"
    class="g-modal"
    id="headlessui-dialog-9"
    role="dialog"
  >
    <div
      class="g-modal__container"
    >
      <div
        aria-hidden="true"
        class="g-modal__overlay"
        id="headlessui-dialog-overlay-10"
      />
      <button
        class="g-modal__test"
      />
      <div
        class="g-modal__content"
      >
        <h2
          class="g-modal__title"
          id="headlessui-dialog-title-11"
        >
          Terms of service
        </h2>
        <p
          id="headlessui-description-12"
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry.
        </p>
        <!---->
        <div
          class="g-modal__footer"
        >
          
          
          <button
            as="button"
            class="g-button"
            fluid="false"
            loading="false"
          >
            <!---->
            <span
              class="g-button__label"
            >
              
              Secondary Action
              
            </span>
          </button>
          <button
            as="button"
            class="g-button g-button--primary"
            fluid="false"
            loading="false"
            variant="primary"
          >
            <!---->
            <span
              class="g-button__label"
            >
              
              Primary Action
              
            </span>
          </button>
          
          
        </div>
      </div>
      <button
        aria-label="Close"
        class="g-modal__close"
        type="button"
      >
        <svg
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
